File: c++-leetcode/LL/LL.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node
{
    Node *next;
    int data;
};

int mid(Node *head)
{
    Node *f = head;
    Node *s = head;
    while (f != NULL && f->next != NULL)
    {
        f = f->next->next;
        s = s->next;
    }
    return s->data;
}

int main()
{
    Node *node = new Node();
    Node *head = node;
    node->data = 2;
    cout << node->data;
    node->next = new Node();
    node = node->next;
    node->data = 4;
    cout << node->data;
    node->next = new Node();
    node = node->next;
    node->data = 6;
    cout << node->data;
    node->next = new Node();
    node = node->next;
    node->data = 8;
    cout << node->data;
    node->next = new Node();
    node = node->next;
    node->data = 10;
    cout << node->data;
    cout << "\n Mid element: " << mid(head);
}

File: c++-leetcode/python/LL.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def revList(self,head):
        curr = head
        prev = None
        next = head.next
        while curr != None:
            curr.next = prev
            prev = curr
            curr = next
            if next != None:
                next = next.next
        return prev
    
    def midNode(self, head):
        s = head
        f = head
        while f != None and f.next != None:
            s = s.next
            f = f.next.next
        return s
    

head = Node('a')
head.next = Node('b')
head.next.next = Node('c')
head.next.next.next = Node('d')

temp = head
while temp != None:
    print(temp.data)
    temp = temp.next

midNode = head.midNode(head)
print(f"mid Node: {midNode.data}")

print(" In rev order")
rHead = head.revList(head)
while rHead != None:
    print(rHead.data)
    rHead = rHead.next